recipe/recipe.go

```diff
@@ -36,3 +42,335 @@ type instance struct {
+		"v4": true,
+	}
+
+	// Validation to check at least one field of the recipe is updated
+	allStringFieldsEmpty := recipe.ID == "" && recipe.Format == "" && recipe.Alias == ""
+	allRemainingFieldsNil := recipe.InputFiles == nil && recipe.OutputInstances == nil
+	if allStringFieldsEmpty && allRemainingFieldsNil {
+		invalidFields = append(invalidFields, "no recipe fields updates given")
+	}
+
+	if recipe.ID != "" {
+		invalidFields = append(invalidFields, "id cannot be changed")
+	}
+	if recipe.Format != "" {
+		if !validFormats[recipe.Format] {
+			invalidFields = append(invalidFields, "format is not valid")
+		}
+	}
+
+	if recipe.InputFiles != nil && len(recipe.InputFiles) > 0 {
+		for i, file := range recipe.InputFiles {
+			if file.Description == "" {
+				missingFields = append(missingFields, "input-files["+strconv.Itoa(i)+"].description")
+			}
+		}
+	}
+	if recipe.InputFiles != nil && len(recipe.InputFiles) == 0 {
+		invalidFields = append(invalidFields, "empty input-files update given")
+	}
+
+	//When doing the update, as recipe.OutputInstances is an array, it needs to make sure that all fields of the instance are complete
+	//This functionality is already available in validateInstance
+	if recipe.OutputInstances != nil && len(recipe.OutputInstances) > 0 {
+		for i, instance := range recipe.OutputInstances {
+			instanceMissingFields, instanceInvalidFields := instance.validateInstance()
+			if len(instanceMissingFields) > 0 {
+				for _, mField := range instanceMissingFields {
+					instanceMissingFields[i] = "output-instances[" + strconv.Itoa(i) + "]." + mField
+				}
+			}
+			if len(instanceInvalidFields) > 0 {
+				for _, iField := range instanceInvalidFields {
+					instanceInvalidFields[i] = "output-instances[" + strconv.Itoa(i) + "]." + iField
+				}
+			}
+			missingFields = append(missingFields, instanceMissingFields...)
+			invalidFields = append(invalidFields, instanceInvalidFields...)
+		}
+	}
+	if recipe.OutputInstances != nil && len(recipe.OutputInstances) == 0 {
+		invalidFields = append(invalidFields, "empty output-instances update given")
+	}
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+
+}
+
+//ValidateUpdateInstance - checks fields of instance before updating the instance of the recipe
+func (instance *Instance) ValidateUpdateInstance() error {
+	var missingFields []string
+	var invalidFields []string
+
+	// Validation to check if at least one instance field is updated
+	allStringFieldsEmpty := instance.DatasetID == "" && instance.Title == ""
+	allRemainingFieldsNil := instance.Editions == nil && instance.CodeLists == nil
+	if allStringFieldsEmpty && allRemainingFieldsNil {
+		invalidFields = append(invalidFields, "no instance fields updates given")
+	}
+
+	if instance.DatasetID != "" {
+		invalidFields = append(invalidFields, "dataset-id cannot be changed")
```

Comment:
Worth running this by Jim - I think on occasion he has changed dataset IDs as he's been working on recipes. Might be quite rare, so potentially correct to be defensive about it